packages/flutter/lib/src/gestures/mouse_tracking.dart

```diff
@@ -104,102 +136,209 @@ class MouseTracker extends ChangeNotifier {
+  void _removeMouseDevice(int device, PointerEvent event) {
+    final bool wasConnected = mouseIsConnected;
+    assert(_mouseStates.containsKey(device));
+    final _MouseState disconnectedMouseState = _mouseStates.remove(device);
+    disconnectedMouseState.mostRecentEvent = event;
+    // Schedule a check to exit annotations that used to contain this pointer.
+    _checkDeviceUpdates(
+      device: device,
+      disconnectedMouseState: disconnectedMouseState,
+    );
+    if (mouseIsConnected != wasConnected) {
+      notifyListeners();
+    }
+  }
+
+  // Handler for events coming from the PointerRouter.
+  void _handleEvent(PointerEvent event) {
+    if (event.kind != PointerDeviceKind.mouse) {
+      return;
+    }
+    final int device = event.device;
+    if (event is PointerAddedEvent) {
+      _addMouseDevice(device, event);
+    } else if (event is PointerRemovedEvent) {
+      _removeMouseDevice(device, event);
+    } else if (event is PointerHoverEvent) {
+      final _MouseState mouseState = _guaranteeMouseState(device, event);
+      final PointerEvent previousEvent = mouseState.mostRecentEvent;
+      mouseState.mostRecentEvent = event;
+      if (previousEvent is PointerAddedEvent || previousEvent.position != event.position) {
+        // Only send notifications if we have our first event, or if the
+        // location of the mouse has changed
+        _checkDeviceUpdates(device: device);
       }
     }
-    _trackedAnnotations.remove(annotation);
   }
 
   bool _scheduledPostFramePositionCheck = false;
-  // Schedules a position check at the end of this frame for those annotations
-  // that have been added.
+  // Schedules a position check at the end of this frame.
+  // It is only called during a frame during which annotations have been added.
   void _scheduleMousePositionCheck() {
     // If we're not tracking anything, then there is no point in registering a
     // frame callback or scheduling a frame. By definition there are no active
     // annotations that need exiting, either.
-    if (_trackedAnnotations.isNotEmpty && !_scheduledPostFramePositionCheck) {
+    if (!_scheduledPostFramePositionCheck) {
       _scheduledPostFramePositionCheck = true;
       SchedulerBinding.instance.addPostFrameCallback((Duration duration) {
-        _sendMouseNotifications(_lastMouseEvent.keys);
+        _checkAllDevicesUpdates();
         _scheduledPostFramePositionCheck = false;
       });
     }
   }
 
-  // Handler for events coming from the PointerRouter.
-  void _handleEvent(PointerEvent event) {
-    if (event.kind != PointerDeviceKind.mouse) {
-      return;
+  // Dispatch callbacks related to a device after all necessary information
+  // has been collected.
+  //
+  // This function should not not change the provided states, and should not
+  // access information that is not provided in parameters (hence being
+  // static).
+  static void _dispatchDeviceCallbacks({
```

Comment:
Maybe move this method after _checkDeviceUpdates. It's only called from there and you kinda have to read that one first to understand what's going on.